src/Storages/StorageKeeperMap.cpp

```diff
@@ -623,6 +640,302 @@ NamesAndTypesList StorageKeeperMap::getVirtuals() const
+            }
+        };
+
+        auto max_multiread_size = with_retries->getKeeperSettings().batch_size_for_keeper_multiread;
+
+        auto keys_it = data_children.begin();
+        while (keys_it != data_children.end())
+        {
+            auto step = std::min(static_cast<UInt64>(std::distance(keys_it, data_children.end())), max_multiread_size);
+            write_rows(std::span{keys_it, keys_it + step});
+            keys_it = keys_it + step;
+        }
+
+        data_out->finalize();
+        data_out.reset();
+        data_out_compressed->finalize();
+        data_out_compressed.reset();
+
+        return {{file_path, std::make_shared<BackupEntryFromAppendOnlyFile>(temp_disk, data_file_path)}};
+    }
+
+    fs::path data_zookeeper_path;
+    DiskPtr temp_disk;
+    std::optional<TemporaryFileOnDisk> temp_dir_owner;
+    UInt64 max_compress_block_size;
+    String file_path;
+    std::shared_ptr<WithRetries> with_retries;
+};
+}
+
+void StorageKeeperMap::backupData(BackupEntriesCollector & backup_entries_collector, const String & data_path_in_backup, const std::optional<ASTs> & /*partitions*/)
+{
+    auto table_id = toString(getStorageID().uuid);
+
+    auto coordination = backup_entries_collector.getBackupCoordination();
+    coordination->addKeeperMapTable(zk_root_path, table_id, data_path_in_backup);
+
+    /// This task will be executed after all tables have registered their root zk path and the coordination is ready to
+    /// assign each path to a single table only.
+    auto post_collecting_task = [my_table_id = std::move(table_id), coordination, &backup_entries_collector, my_data_path_in_backup = data_path_in_backup, this]
+    {
+        auto path_with_data = coordination->getKeeperMapDataPath(zk_root_path);
+        if (path_with_data == my_data_path_in_backup)
+        {
+            auto temp_disk = backup_entries_collector.getContext()->getGlobalTemporaryVolume()->getDisk(0);
+            auto max_compress_block_size = backup_entries_collector.getContext()->getSettingsRef().max_compress_block_size;
+
+            auto with_retries = std::make_shared<WithRetries>
+            (
+                &Poco::Logger::get(fmt::format("StorageKeeperMapBackup ({})", getStorageID().getNameForLogs())),
+                [&] { return getClient(); },
+                WithRetries::KeeperSettings::fromContext(backup_entries_collector.getContext()),
+                [](WithRetries::FaultyKeeper &) {}
+            );
+
+            backup_entries_collector.addBackupEntries(
+                std::make_shared<KeeperMapBackup>(this->zk_data_path, path_with_data, temp_disk, max_compress_block_size, std::move(with_retries))
+                    ->getBackupEntries());
+            return;
+        }
+
+        auto file_path = fs::path(my_data_path_in_backup) / backup_data_location_filename;
+        backup_entries_collector.addBackupEntries({{file_path, std::make_shared<BackupEntryFromMemory>(path_with_data)}});
```

Comment:
I thought about this idea of using a backup entry which is actually a path to another place in a backup. I like the idea, probably we could use it in some other storages as well. And we could make it more general with changes in the code like the following:

```
class IBackupEntry
{
public:
   ...
   virtual bool isReference() const { return false; }
   virtual getReferenceTarget() const { throw Exception(NOT_IMPLEMENTED); }
};


/// Represents a reference to another backup entry.
class BackupEntryReference : public IBackupEntry
{
public:

    bool isReference() const override { return true; }
    String getReferenceTarget() const override { return reference_target; }

private:
    String reference_target;
};


struct BackupFileInfo
{
    ...
    
    std::optional<String> reference_target;
};


BackupFileInfo buildFileInfoForBackupEntry(...)
{
    BackupFileInfo info;
    info.file_name = adjusted_path;

    if (backup_entry->isReference())
    {
        info.reference_target = backup_entry->getReferenceTarget();
        return info;
    }
    
    ...
}


void BackupCoordinationFileInfos::prepare() const
{
    std::sort(file_infos_for_all_hosts ...)
    adjancent_find(...)
    ...
    
    /// Assign sizes and checksums for all the references.
    for (auto * file_info : file_infos_for_all_hosts)
    {
        if (file_info->reference_target)
        {
            auto * ref_file_info = std::find(...)
            file_info->size = ref_file_info->size;
            file_info->checkrum = ref_file_info->checksum;
            file_info->base_size = ref_file_info->size;
            file_info->base_checkrum = ref_file_info->checksum;
            file_info->size = ref_file_info->size;
            file_info->checkrum = ref_file_info->checksum;
            file_info->encrypted_by_disk = ref_file_info->encrypted_by_disk;
        }
    }
    
    ...
}
```

Then instead of
```
backup_entries_collector.addBackupEntries({{file_path, std::make_shared<BackupEntryFromMemory>(path_with_data)}});
```
we could write
```
backup_entries_collector.addBackupEntries({{file_path, std::make_shared<BackupEntryReference>(path_with_data)}});
```
and then `restoreDataFromBackup()` won't have to do anything about `data_location.bin` at all.

Reply:
This was a really nice idea so I implemented it.
It's a bit tricky because there are a lot of places where file_info is used but I think I covered all relevant.
PTAL

Reply:
Thank you!